File: 04._basic_servers/data_format_translation_servers/node/parsers.js
import fs from "fs/promises";
import { parse } from "csv-parse/sync";
import { XMLParser } from "fast-xml-parser";
import { parse as YAMLparse } from "yaml";

export async function parseCSV(path) {
  const data = await fs.readFile(path, "utf-8");
  const parsedData = parse(data, {
    delimiter: ",",
    columns: true,
  });
  return parsedData;
}

export async function parseJSON(path) {
  return await import(path, { assert: { type: "json" } });
}

export async function parseXML(path) {
  const data = await fs.readFile(path, "utf-8");
  const parser = new XMLParser();
  const parsedData = parser.parse(data);
  return parsedData;
}

export async function parseYAML(path) {
  const data = await fs.readFile(path, "utf-8");
  const parsedData = YAMLparse(data);
  return parsedData;
}

export async function parseTXT(path) {
  const data = await fs.readFile(path, "utf-8");
  const obj = {};
  const lines = data.split("\r\n");
  lines.forEach((line) => {
    const lineArray = line.split(" ");
    switch (lineArray[1]) {
      case "@string":
        obj[lineArray[0]] = lineArray[2];
        break;
      case "@number":
        obj[lineArray[0]] = Number(lineArray[2]);
        break;
      case "@array":
        obj[lineArray[0]] = lineArray[2].split(",");
        break;
    }
  });
  return obj;
}

File: 04._basic_servers/data_format_translation_servers/python/parsers/parsers.py
import csv
import json
import xmltodict
import yaml


def parseCSV(path):
    with open(path) as csvfile:
        reader = csv.DictReader(csvfile)
        rows = []
        for row in reader:
            rows.append(row)
        return rows


def parseJSON(path):
    with open(path) as jsonfile:
        reader = json.load(jsonfile)
        rows = []
        for row in reader:
            rows.append(row)
        return rows


def parseXML(path):
    with open(path) as xmlfile:
        data = xmltodict.parse(xmlfile.read())
        return data


def parseYAML(path):
    with open(path) as yamlfile:
        data = yaml.safe_load(yamlfile)
        return data


def parseTXT(path):
    dict = {}
    with open(path) as txtfile:
        for line in txtfile:
            parseLine(line, dict)

    return dict


def parseLine(line: str, dict):
    key, tag, value = line.split()
    match tag:
        case "@string":
            dict[key] = value
        case "@number":
            dict[key] = int(value)
        case "@array":
            dict[key] = value.split(",")
